src/fuse_scan.cpp

```diff
@@ -7,15 +7,96 @@
 #include <laser_geometry/laser_geometry.h>
 #include <vector>
 #include <algorithm>
-#include <math.h>
+#include <cmath>
+#include <limits>
 
+#define EPSI 10e-6
+
+class Point {
+  public:
+  Point(double _x, double _y) {
+      x = _x;
+      y = _y;
+  }
+
+  double x;
+  double y;
+};
+
+bool onSegment(Point p, Point q, Point r) { 
+    if (q.x <= std::max(p.x, r.x) && q.x >= std::min(p.x, r.x) && 
+            q.y <= std::max(p.y, r.y) && q.y >= std::min(p.y, r.y)) 
+        return true; 
+    return false; 
+}
+
+int orientation(Point p, Point q, Point r) { 
+    int val = (q.y - p.y) * (r.x - q.x) - 
+            (q.x - p.x) * (r.y - q.y); 
+ 
+    if (val <= EPSI)
+      return 0;
+    return (val > 0)? 1: 2;
+}
+
+bool doIntersect(Point p1, Point q1, Point p2, Point q2) { 
+    int o1 = orientation(p1, q1, p2); 
+    int o2 = orientation(p1, q1, q2); 
+    int o3 = orientation(p2, q2, p1); 
+    int o4 = orientation(p2, q2, q1); 
+ 
+    // General case 
+    if (o1 != o2 && o3 != o4) 
+        return true; 
+ 
+    // Special Cases 
+    // p1, q1 and p2 are colinear and p2 lies on segment p1q1 
+    if (o1 == 0 && onSegment(p1, p2, q1))
+      return true; 
+ 
+    // p1, q1 and p2 are colinear and q2 lies on segment p1q1 
+    if (o2 == 0 && onSegment(p1, q2, q1))
+      return true; 
+ 
+    // p2, q2 and p1 are colinear and p1 lies on segment p2q2 
+    if (o3 == 0 && onSegment(p2, p1, q2))
+      return true; 
+ 
+    // p2, q2 and q1 are colinear and q1 lies on segment p2q2 
+    if (o4 == 0 && onSegment(p2, q1, q2))
+      return true; 
+ 
+    return false;
+} 
+
+bool isInside(std::vector<Point> &polygon, Point p) {
+    if (polygon.size() < 3) return false; 
+ 
+    Point extreme = {std::numeric_limits<double>::max(), p.y}; 
+
+    int count = 0, i = 0; 
+    do { 
+        int next = (i+1)%polygon.size(); 
+
+        if (doIntersect(polygon[i], polygon[next], p, extreme)) {
+            if (orientation(polygon[i], p, polygon[next]) == 0) 
+            return onSegment(polygon[i], p, polygon[next]); 
```

Comment:
is it basically finding if the points are on the edges of the polygon?

I need to understand the algo here. Will call you

also, doesn't pcl provide a filter for this?

Reply:
i found this
https://pointclouds.org/documentation/classpcl_1_1_crop_hull.html